plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSource.java

```diff
@@ -132,140 +150,83 @@ public long getReadTimeNanos()
     }
 
     @Override
-    public long getCompletedBytes()
+    public boolean isFinished()
     {
-        return 0;
+        return finished;
     }
 
     @Override
-    public Type getType(int field)
+    public Page getNextPage()
     {
-        return columnHandles[field].getColumnType();
-    }
-
-    @Override
-    public boolean advanceNextPosition()
-    {
-        if (closed) {
-            return false;
-        }
-
+        verify(pageBuilder.isEmpty());
         try {
             if (resultSet == null) {
-                long start = System.nanoTime();
-                Future<ResultSet> resultSetFuture = executor.submit(() -> {
-                    log.debug("Executing: %s", statement);
-                    return statement.executeQuery();
-                });
-                try {
-                    // statement.executeQuery() may block uninterruptedly, using async way so we are able to cancel remote query
-                    // See javadoc of java.sql.Connection.setNetworkTimeout
-                    resultSet = resultSetFuture.get();
-                }
-                catch (ExecutionException e) {
-                    if (e.getCause() instanceof SQLException cause) {
-                        SQLException sqlException = new SQLException(cause.getMessage(), cause.getSQLState(), cause.getErrorCode(), e);
-                        if (cause.getNextException() != null) {
-                            sqlException.setNextException(cause.getNextException());
-                        }
-                        throw sqlException;
+                resultSet = getFutureValue(resultSetFuture);
+                verify(resultSet != null, "resultSet is null");
```

Comment:
`resultSet = requireNonNull(getFutureValue(resultSetFuture), "resultSet is null")`